purchase_request_to_rfq/models/purchase_order.py

```diff
@@ -75,6 +76,71 @@ def button_confirm(self):
+        """
+        po_from_request = self.filtered(
+            lambda po: po.mapped('order_line.purchase_request_lines'))
+        remaining_po = self.filtered(
+            lambda po: not po.mapped('order_line.purchase_request_lines'))
+        po_from_request.cancel_po_from_request()
+        return super(PurchaseOrder, remaining_po).button_cancel()
+
+    @api.multi
+    def cancel_po_from_request(self):
+        """Filter PO lines of PO that are related to some purchase request
+        (PR).
+        * Lines which have no relation to PRs are processed with a copy of
+        standard code.
+        * Lines with relation to PRs aren't processed. The purpose of this is
+        to leave on hands of the purchase requests the task of cancelling
+        procurements which originated them."""
+        for order in self:
+            for pick in order.picking_ids:
+                if pick.state == 'done':
+                    raise UserError(_(
+                        'Unable to cancel purchase order %s as some '
+                        'receptions have already been done.') % order.name)
+            for inv in order.invoice_ids:
+                if inv and inv.state not in ('cancel', 'draft'):
+                    raise UserError(_(
+                        "Unable to cancel this purchase order. You must "
+                        "first cancel related vendor bills."))
+            for pick in order.picking_ids.filtered(
+                    lambda r: r.state != 'cancel'):
+                pick.action_cancel()
+            # Post a msg in purchase requests:
+            request_po_lines = order.order_line.filtered(
+                lambda pol: pol.purchase_request_lines)
+            for line in request_po_lines:
+                message = \
+                    line._po_line_purchase_request_cancel_message_content()
+                requests = line.purchase_request_lines.mapped('request_id')
+                for req in requests:
+                    req.message_post(body=message, subtype='mail.mt_comment')
+            # Search for lines not created from a Purchase request and end
+            # the process
+            no_request_po_lines = order.order_line.filtered(
+                lambda pol: not pol.purchase_request_lines)
+            # TODO: remove PO from procurement ?!
+            if not self.env.context.get('cancel_procurement'):
+                procurements = no_request_po_lines.mapped('procurement_ids')
+                procurements.filtered(lambda r: r.state not in (
+                    'cancel', 'exception') and r.rule_id.propagate).write(
+                    {'state': 'cancel'})
+                procurements.filtered(lambda r: r.state not in (
+                    'cancel', 'exception') and not r.rule_id.propagate).write(
+                    {'state': 'exception'})
+                moves = procurements.filtered(
+                    lambda r: r.rule_id.propagate).mapped('move_dest_id')
```

Comment:
You loop 3 times on ```procurements```. I don't know if the performance loss is significative. What do you think @sbidoul?